packages/cli/src/containers/StepRunner.tsx

```diff
@@ -0,0 +1,168 @@
+  error: (stepOutput: { error: Error }) => React.ReactNode;
+};
+
+export type StepRunnerActionOptions<
+  S extends Record<string, StepOutput> = Record<string, StepOutput>
+> = {
+  config: Config;
+  stepOutputs: S;
+};
+
+export type StepRunnerStep<O extends StepOutput = StepOutput> = {
+  name: string;
+  emoji: EmojiCode;
+  messages: StepMessages<O>;
+  action: (options: StepRunnerActionOptions) => Promise<O>;
+};
+
+type StepStatusReducerAction = {
+  step: number;
+  state: StepState;
+};
+
+function stepStatusReducer(state: StepState[], action: StepStatusReducerAction) {
+  const newState = [...state];
+  newState[action.step] = action.state;
+  return newState;
+}
+
+type StepOutputReducerAction = {
+  stepName: string;
+  output: StepOutput;
+};
+
+function stepOutputReducer(state: Record<string, StepOutput>, action: StepOutputReducerAction) {
+  return {
+    ...state,
+    [action.stepName]: action.output,
+  };
+}
+
+export type OnSuccessOptions<O = any> = {
+  stepOutputs: O;
+  startTime: Date;
+  endTime: Date;
+  runtimeInSeconds: number;
+};
+
+export type StepRunnerProps<O = any> = {
+  steps: StepRunnerStep[];
+  onSuccess?: (options: OnSuccessOptions<O>) => void;
+};
+
+export function StepRunner({ steps, onSuccess }: StepRunnerProps) {
+  const config = useConfig();
+  const [currentStep, setCurrentStep] = React.useState<number>(0);
+  const [stepStatuses, setStepStatus] = React.useReducer(stepStatusReducer, []);
+  const [stepOutputs, setStepOutput] = React.useReducer(stepOutputReducer, {});
+
+  const [error, setError] = React.useState<Error>();
+  const [done, setDone] = React.useState(false);
+  const [startTime, setStartTime] = React.useState<Date>();
+  const [endTime, setEndTime] = React.useState<Date>();
+
+  React.useEffect(() => {
+    setStartTime(new Date());
+  }, []);
+
+  const runCurrentStep = React.useCallback(async () => {
+    setStepStatus({ step: currentStep, state: StepState.RUNNING });
+
+    try {
+      const output = await steps[currentStep].action({
+        config,
+        stepOutputs,
+      });
+      setStepOutput({
+        stepName: steps[currentStep].name,
+        output,
+      });
+      setStepStatus({ step: currentStep, state: StepState.SUCCESS });
+      setCurrentStep(currentStep + 1);
+    } catch (err) {
+      setStepStatus({ step: currentStep, state: StepState.FAILED });
+      setStepOutput({
+        stepName: steps[currentStep].name,
+        output: { error: err },
+      });
+      setError(err);
+    }
+  }, [config, currentStep, steps, stepOutputs]);
+
+  React.useEffect(() => {
+    // If current step has passed passed last step OR step is already started OR on error
+    // -> don't start step
+    if (currentStep === steps.length || stepStatuses[currentStep] || error) return;
+    runCurrentStep();
+  }, [steps, error, currentStep, stepStatuses, runCurrentStep]);
+
+  React.useEffect(() => {
+    if (currentStep === steps.length && !done) {
```

Comment:
Problem is, that the `currentStep` at this point is already bigger than the last step index.
In my opinion it's good to assume that it's either passed or failed when the steps have surpassed that index, because the `currentStep` is only increased on success or failure

Reply:
ok